rust/datafusion/src/datasource/parquet.rs

```diff
@@ -0,0 +1,688 @@
+use parquet::data_type::Int96;
+
+pub struct ParquetTable {
+    filename: String,
+    schema: Arc<Schema>,
+}
+
+impl ParquetTable {
+    pub fn new(filename: &str) -> Self {
+        let file = File::open(filename).unwrap();
+        let parquet_file = ParquetFile::open(file, None).unwrap();
+        let schema = parquet_file.schema.clone();
+        Self {
+            filename: filename.to_string(),
+            schema,
+        }
+    }
+}
+
+impl Table for ParquetTable {
+    fn schema(&self) -> &Arc<Schema> {
+        &self.schema
+    }
+
+    fn scan(
+        &self,
+        projection: &Option<Vec<usize>>,
+        _batch_size: usize,
+    ) -> Result<Vec<ScanResult>> {
+        let file = File::open(self.filename.clone()).unwrap();
+        let parquet_file = ParquetFile::open(file, projection.clone()).unwrap();
+        Ok(vec![Arc::new(Mutex::new(parquet_file))])
+    }
+}
+
+pub struct ParquetFile {
+    reader: SerializedFileReader<File>,
+    row_group_index: usize,
+    /// The schema of the underlying file
+    schema: Arc<Schema>,
+    projection: Vec<usize>,
+    batch_size: usize,
+    current_row_group: Option<Box<RowGroupReader>>,
+    column_readers: Vec<ColumnReader>,
+}
+
+impl ParquetFile {
+    pub fn open(file: File, projection: Option<Vec<usize>>) -> Result<Self> {
+        println!("open()");
+
+        let reader = SerializedFileReader::new(file).unwrap();
+
+        let metadata = reader.metadata();
+        let file_type = to_arrow(metadata.file_metadata().schema())?;
+
+        match file_type.data_type() {
+            DataType::Struct(fields) => {
+                let schema = Schema::new(fields.clone());
+                //println!("Parquet schema: {:?}", schema);
+
+                let projection = match projection {
+                    Some(p) => p,
+                    None => {
+                        let mut p = Vec::with_capacity(schema.fields().len());
+                        for i in 0..schema.fields().len() {
+                            p.push(i);
+                        }
+                        p
+                    }
+                };
+
+                let projected_fields: Vec<Field> = projection
+                    .iter()
+                    .map(|i| schema.fields()[*i].clone())
+                    .collect();
+
+                let projected_schema = Arc::new(Schema::new(projected_fields));
+
+                Ok(ParquetFile {
+                    reader: reader,
+                    row_group_index: 0,
+                    schema: projected_schema,
+                    projection,
+                    batch_size: 64 * 1024,
+                    current_row_group: None,
+                    column_readers: vec![],
+                })
+            }
+            _ => Err(ExecutionError::General(
+                "Failed to read Parquet schema".to_string(),
+            )),
+        }
+    }
+
+    fn load_next_row_group(&mut self) {
```

Comment:
Maybe we should return a `Result` from this function? also in the `else` branch we should not just `panic`.